リアクションを提供するカスタムフックの実装

sts11142 opened the pull request:
resolve: #6


## やったこと

- リアクションの型を定義
- 全リアクションを提供するカスタムフックを実装
- フックの呼び出し例を実装

## 補足

- ポジティブネガティブで右に左に発射したりする時に使うと思って`sentiment`属性も追加してます
- `type`属性は一旦CSSで横幅を切り分けるために定義しました
- あとベースのHTMLとかスタイルをいじりました 🙏 都合悪いところあれば教えてください！（styled-component詳しくなく...）

sts11142 committed:
feat: reactionを提供するカスタムフックを定義

呼び出し側のCSS制御を鑑みて`type`を導入（`Reaction['text']`の要素の幅が異なるため）

sts11142 committed:
feat: リアクションの適用サンプルを実装

sts11142 committed:
refactor: 属性値の名前をより分かりやすく

zackerms commented on frontend/src/hooks/use-reaction.ts:
関数に合わせて、`useReaction.ts`にしたいです！

zackerms commented on frontend/src/hooks/use-reaction.ts:
```diff
@@ -0,0 +1,58 @@
+import { useMemo } from 'react';
+
+type Reaction = {
+  type: 'emoji' | 'phrase';
```
[refactor] こんな感じで書くと、変更に強くなりそう

```ts
export const ReactionType = {
  EMOJI: 'emoji',
  PHRASE: 'phrase',
};
export type ReactionType = typeof ReactionType[keyof typeof ReactionType];
```

```ts
type: ReactionType.PHRASE
```

sts11142 replied:
なるほど！こうすると変更にも強くなるしプロパティの値自体も型として他で流用できるね 👍 
勉強になります！

zackerms reviewed (commented):
@sts11142 コメントしましたー

sts11142 committed:
chore: rename

sts11142 committed:
refactor: 変更容易性を考慮してReactionのプロパティを定数ベースの記述に変更

型exportは必要なタイミングでよしなにどうぞ

zackerms commented on frontend/src/hooks/useReaction.ts:
```diff
@@ -0,0 +1,79 @@
+import { useMemo } from 'react';
+
+const ReactionType = {
+  EMOJI: 'emoji',
+  PHRASE: 'phrase',
+} as const;
+type ReactionType = (typeof ReactionType)[keyof typeof ReactionType];
+
+const ReactionText = {
+  EMOJI_SET: ['🤯', '😑', '🤔', '👍', '🥹', '🤩'] as const,
+  PHRASE_SET: ['わかる', 'わからん'] as const,
+};
+type EmojiText = (typeof ReactionText.EMOJI_SET)[number];
+type PhraseText = (typeof ReactionText.PHRASE_SET)[number];
+type ReactionTextType = EmojiText | PhraseText;
+
+const SentimentType = {
+  POSITIVE: 'positive',
+  NEGATIVE: 'negative',
+} as const;
+type SentimentType = (typeof SentimentType)[keyof typeof SentimentType];
+
+type Reaction = {
+  type: ReactionType;
+  text: ReactionTextType;
+  sentiment: SentimentType;
+};
```
@sts11142 いい感じ！ このhooksがなくなっても
型は影響を受けないようにするために `src/models/Reaction.ts`
 みたいなファイルを作って切り出すと良さそう！

zackerms replied:
@sts11142 あとあと考えてみたんだけど、サーバから受信するメッセージでは`type`とかのプロパティが無いから、
```ts
type Message { 
   message: string 
}

if (positiveEmojis.includes(message.message)) {
  
}
```
みたいに、if文で分けるっていう方法でもいいのかなと思いました。

zackerms commented on frontend/src/pages/rooms/[id]/index.tsx:
```diff
@@ -116,19 +116,16 @@ const MessageSendButton = styled.button`
+    cursor: pointer;
     width: 48px;
     height: 48px;
+    padding: 2px;
```
@sts11142 このpaddingってなんのためにあるんだっけ 🤔

sts11142 replied:
少し余白加えたら可愛くなったっていう笑
これは諸説あり...

zackerms reviewed (commented):
@sts11142 追加でコメントしましたー

sts11142 committed:
Merge branch 'main' into feature/reaction-customhook

sts11142 committed:
feat: websocketで受け取るmessageの形に合わせてデータを再定義（叩き）

Message型に合わせる形で簡潔にリアクションを扱いたい

sts11142 committed:
delete: 一旦不要ぽいカスタムフック

メモ：絵文字かどうか、センチメンタルの属性は、みたいな内容はリアクションテキストの値から分かりそう→Message型に従う形で、必要な定義をModelに記述する方針に変更

sts11142 commented:
@zackerms 
とりあえずこんなイメージ？というのを書いてみた！
また意見ください 🙏

sts11142 committed:
Merge branch 'main' into feature/reaction-customhook

sts11142 committed:
refactor: `emoji`は包括的に`reaction`に置き換えた方が良さそう？ & イベントハンドラって分かるようにhandleって名付けた方が良さそう

zackerms reviewed (approved):
![LGTM](https://image.lgtmoon.dev/272744)
